Shard the public llvm-config.h in multiple files (NFC)

joker-eph opened the pull request:
This will allow for more granular include and improve incremental rebuild and caching when switching targets or other options independently.

A single change of a target, default triple, or any option is triggering a monolithic rebuild of any user of a single of these options. We should make it as granular as possible.

As a follow-up we'll update the user of llvm-config.h incrementally.
Another follow-up should be to look into the private config.h the same way.

joker-eph committed:
Shard the public llvm-config.h in three files (NFC)

This will allow for more granular include and improve incremental
rebuild and caching when switching targets or other options independently.

A single change of a target, default triple, or any option is
triggering a monolithic rebuild of any user of a single of these
options. We should make it as granular as possible.

There are also option in llvm-config.h that seems to belong to
config.h instead (that is it's not clear why they are part of the
public LLVM distribution). For example the aspects relative to the
"native" targets have to with the build machine and in a
cross-compiler environment shouldn't be in the install package.

Other candidates includes LLVM_ENABLE_DIA_SDK, HAVE_SYSEXITS_H,
LLVM_ON_UNIX.

fabianmcg commented on llvm/include/llvm/Config/llvm-config-targets.h.cmake:
```diff
@@ -0,0 +1,121 @@
+/* Define if the Xtensa target is built in */
+#cmakedefine01 LLVM_HAS_XTENSA_TARGET
+
+#endif // LLVM_CONFIG_TARGETS_H
```
nit, new line at end of file.

joker-eph replied:
```suggestion
#endif // LLVM_CONFIG_TARGETS_H

```

fabianmcg reviewed (approved):
LGTM! And I can confirm that enabling or disabling targets has a minimal rebuild impact if Ccache is enabled.

joker-eph commented:
> I can confirm that enabling or disabling targets has a minimal rebuild impact if Ccache is enabled.

Something fishy: this change shouldn't change anything until we update users.

fabianmcg commented:
> Something fishy: this change shouldn't change anything until we update users.

Switching on & off the target does prompt ninja to rebuild everything, however, if a normal build takes in my machine 10 min, the rebuild of all targets took 2 min with Ccache.

joker-eph commented:
Right but that should be the case identically with and without this patch I think: do you see anything different?

fabianmcg commented:
I just cleared the cache, cleaned the tree and retried with a clean build and you're right.

joker-eph commented:
@nico @fabianmcg  : I ended up biting the bullet and do an almost full sharding of the file, PTAL.

I also removed the new options I introduced yesterday for testing is a target is enabled from the llvm-config.h: since this file only exists for backward compatiblity now and these options haven't been widely adopted.

arichardson commented on llvm/include/llvm/Config/llvm-config-enable-curl.h.cmake:
```diff
@@ -0,0 +1,16 @@
+#define LLVM_CONFIG_ENABLE_CURL_H
+
+/* Define if we have curl and want to use it */
+#cmakedefine LLVM_ENABLE_CURL ${LLVM_ENABLE_CURL}
```
I worry that this split is going to cause subtle bugs if we don't use cmakedefine01 everywhere. Otherwise we could end up with #ifndef being evaluated incorrectly.

joker-eph replied:
At the moment we keep including everything through the llvm-config.h ; but you're right this is a major hurdle.

I'm already going one by one through all the options and converting them to cmakedefine01.
I am also making the LLVM codebase `-Wundef` clean: this is the only way to catch this in a principle way (I think we should enable -Wundef by default actually).

joker-eph replied:
I had a long flight yesterday, I have locally all the patches for making every option a `cmakedefine01` and making the codebase -Wundef clean.

nico commented:
Having one file per setting for settings toggled by autoconf (`HAVE_SYSEXITS_H` etc) doesn't seem terribly useful – those aren't user-settable and can just stay in the main file, right?

joker-eph commented:
> If we do this, what's the advantage of putting all these headers in Config/? LLVM_ENABLE_DIA_SDK for example is used exclusively in llvm/tools/llvm-symbolizer/llvm-symbolizer.cpp – should we put the header for that over there then?

In Config/ there is `llvm-config.h` and `config.h`: the former is for things that are used in our public headers or must be made available publicly (through our `install/llvm/include/`). The latter is for private things, and I actually had written initially in this PR description that some of these should move to be private instead: it may be the case for LLVM_ENABLE_DIA_SDK for example. 
However I removed it from the description because I am not totally sure: something that is used only in a tool can be defined in CMake and never in a header. Right now I can't know if a setting is here "accidentally" (someone thought they were "doing the right thing" by adding a setting in the header) or meant to be exposed to third-party in the distribution (that is: even without any in-tree user for a setting, it may be a useful setting for external user to guard code depending on how LLVM was configured, I'm actually building my downstream project that way against a prebuilt LLVM and we support different configs). 
So that requires a bit more archeology and consideration on a case by case. (I can also just make it "private" and wait for someone to complain...).

> I think this is a noble patch, but doing this piecemeal for the ones people actually toggle is imho good enough.

I started to do that and I realized that there were only a couple of options like that (HAVE_SYSEXIT and HAS_ATOMICS I think), I can group them into "llvm-config-platform.h.cmake" if you prefer. Right now `llvm/include/llvm/Support/ExitCodes.h` is the only user of HAVE_SYSEXITS_H, I was seeing it as some kind of "private config" that is here only because it's used in a public header.
There is also `llvm-config-version.h.cmake`  but seemed to me preferable to not have version included everywhere has_atomics is ("platforms" flag are also conceptually different than LLVM-version).

> (I would love to be able to turn assertions on and off without a full rebuild, but even if that's in its own .h, assertions are used anywhere, so it's still a full rebuild even then :P)

(right it is affecting virtually everything, even then we don't define NDEBUG ourselves anyway, and we don't have a macro for our own assertions I believe)

llvm-beanz commented:
I worry about the tradeoff here. How much does this really speed up builds with a caching layer when you change these settings?

I feel like in a lot of cases, the caching layer should already be resilient against macros that aren't used because the caches tend to identify collisions based on preprocessed files, so if you don't use a macro you should still get a cache hit. This might bypass cache lookups, but it comes at a cost to maintainability.

joker-eph commented:
What is the maintainability concern? I don’t quite get what you have in mind on this.
You include what you use (clang-tidy has a header include cleanup helping with this), and Wundef catches any issue.

llvm-beanz commented:
Adding new files for new options and splitting defines across a bunch of files is messy and will incur maintenance as people have a whole bunch of files floating around. Which isn't to say we shouldn't do this, but have you benchmarked how much this can improve build times? This is a lot of complexity to add to our includes and build settings system if it doesn't have a significant benefit.

joker-eph commented:
> Adding new files for new options and splitting defines across a bunch of files is messy and will incur maintenance as people have a whole bunch of files floating around.

I don't quite get what you mean: what are "floating" files? I'm trying to picture it, but it's very fuzzy, I prompted StableDiffusion to get help ;)

![download](https://github.com/llvm/llvm-project/assets/3372300/de09ef3c-aca0-4cbd-a0ef-286570306cc0)

> This is a lot of complexity

Sorry but I see zero added complexity really right now. The whole thing seems just to me to be a cleaner software structure where you include what you need (include llvm-config-version.h for the VERSION* macros). Not clear why the current "monolithic blog" would be less messy and less complex?

arsenm reviewed (commented):
I think using cmakedefine01 is an improvement. As far as splitting every option into its own header, I've never wanted to dynamically flip between these. These mostly look like the kind of options I'd be surprised would even work to change after setting once in the CMakeCache. 

Isn't there an additional cost to reading N smaller files?

petrhosek commented:
When debugging builds, I sometimes look at `llvm-config.h` to see what options were set. With this change that becomes basically impossible because I'd need to look at 50 different files.

joker-eph commented:
> I think the important point here is that you need to justify why this is an improvement

Well I am claiming this is a cleanup: which is subjective, but just as much as a lot of software engineer is (hey you started with a lot of "subjective" claims!). Can make it less subjective by claiming separation of concerns and modularity: which is actually measurable. 

 > Do you have performance numbers to show that this speeds up builds in some configuration? 
 
Note that this patch alone does not change anything: because I need to update all the users first and stop including llvm-config.h, which is a series of follow-up cleanup patches.

But the impacts seemed obvious to me, let me actually give numbers, it is fairly easily collect them:

 - There are 25 files which uses LLVM_VERSION_*, yet the LLVM_VERSION_ macro impacts all the files which transitively include llvm-config.h (that is ~ all of the project?). So including LLVM_VERSION where it's used will reduce by O(100-1000) the number of compilation commands affected.
 - LLVM_ENABLE_CURL is used in 2 files, but the macro impacts all the files including `llvm-config.h` transitively, so that is again a factor of O(1000) reduction. 
 - ... (I can continue to grep, but think it'll be the same for all of these?)
 
For example, switching the LLVM_ENABLE_CURL ON/OFF config, on one build config (not including flang/clang) it means 5126 targets to rebuild before, and only 20 after (this is what I observe with `ninja` locally in practice).

llvm-beanz commented:
> But the impacts seemed obvious to me, let me actually give numbers, it is fairly easily collect them:
> 
> * There are 25 files which uses LLVM_VERSION_*, yet the LLVM_VERSION_ macro impacts all the files which transitively include llvm-config.h (that is ~ all of the project?). So including LLVM_VERSION where it's used will reduce by O(100-1000) the number of compilation commands affected.
> * LLVM_ENABLE_CURL is used in 2 files, but the macro impacts all the files including `llvm-config.h` transitively, so that is again a factor of O(1000) reduction.
> * ... (I can continue to grep, but think it'll be the same for all of these?)
> 
> For example, switching the LLVM_ENABLE_CURL ON/OFF config, on one build config (not including flang/clang) it means 5126 targets to rebuild before, and only 20 after (this is what I observe with `ninja` locally in practice).

Your initial comment referred to caching. If you apply `ccache` or `scache`, both preprocess before determining a cache collision, which means that extremely few things rebuild. With object file caching utilities you don't get penalized for preprocessor defines that aren't used in a given translation unit, so it seems that to me that the benefits would be very small.

This continues to lead me to the conclusion that I'm not convinced this change provides enough benefit to justify complicating our configuration include paradigm. Having a single (or small number) of headers that are autoconf or cmake generated is fairly common, having one header per configuration setting is not something I've ever seen a project do.

petrhosek commented:
`LLVM_VERSION_*` changes every 6 months, I don't think that's something we should be optimizing for (but I'd be still fine moving it to `llvm-version.h` if you think it's worth it). For all other options like `LLVM_ENABLE_CURL`, to change the value you need to reconfigure the build, is that a scenario that developers do often enough that we need to optimize for?

joker-eph commented:
> Your initial comment referred to caching.

No: I wrote "incremental rebuild **and** caching". 

> if you apply ccache or scache, both preprocess before determining a cache collision,

Both have support for "direct mode" where preprocessing can be bypassed, the impact can be going from 7 to 11min (or 60s to 110s), see discussion here: https://github.com/llvm/llvm-zorg/pull/68#issuecomment-1807514000

> This continues to lead me to the conclusion that I'm not convinced this change provides enough benefit to justify complicating our configuration include paradigm. 

I don't see the "complicating" aspect really, but also I'm a bit annoyed by your seemingly goal shifting approach to the review and tactic of being vague and evasive on your comments to ultimately get me to waste time on experiment that were obvious. Please don't do that, this isn't productive.

MaskRay commented:
I prefer not to engage in the debate and hope we can focus on the technical aspects instead.

Splitting a config header into multiple files strikes me as somewhat unusual.

I don't frequently update `LLVM_TARGETS_TO_BUILD`, but I do occasionally update `LLVM_EXPERIMENTAL_TARGETS_TO_BUILD`.
I typically build all experimental targets to avoid unnecessarily causing issues with my changes.
However, some experimental targets might encounter random failures, prompting me me to remove them from the list. (Sometimes I try to fix them instead).
I'll be happy if I don't need to rebuild all targets if I remove a single entry from `LLVM_EXPERIMENTAL_TARGETS_TO_BUILD`.

As for other configurations like curl,zstd,threads,etc, I rarely make updates, and I am okay with the rebuild time if I do decide to change them.

nico commented:
To make progress here, can we please just put the macros added in #71164 into their own header (which can live in mlir/, the only consumer of these values) and not do the rest for now?

joker-eph commented:
Sorry but I don't think what you're proposing is a good path forward. And I strongly object to making this an MLIR specific thing: it's not.

llvm-beanz commented:
> Sorry but I don't think what you're proposing is a good path forward. And I strongly object to making this an MLIR specific thing: it's not.

If MLIR is the only consumer of those values reducing the exposure area of them seems like a good idea.

I didn't see the earlier PR go in, but I don't think those values should have been exposed to LLVM in the first place. Exposing target availability preprocessor macros is a significant shift for LLVM and could lead to violating LLVM's library layering. LLVM has a longstanding convention that code dependent on a target being enabled is contained to the target itself.

It seems to me that sharding the configuration files as this PR does isn't something the community has a clear agreement on. Further, the people with the most experience maintaining the LLVM build system (which this change impacts) are voicing opposition.
